### PR TITLE
ChromeOS/Android trackpad gestures

### PR DESCRIPTION
**Pull request by moffatman**
Start sending PointerPanZoom events from ChromeOS/Android when trackpad is used

Fixes https://github.com/flutter/flutter/issues/23604

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Review by blasten — CHANGES_REQUESTED** [review]
Thanks for the contribution! It will really help to get some unit tests and comments for each case that this PR is handling.

**Comment by moffatman** [comment]
@chinmaygarde Framework side needs to be merged first so I haven't spent any time on adding tests here yet (also this platform is the least priority, trackpad gestures already functionally work). Should get around to it within the next week.

**Comment by justinmc** [comment]
If the framework PR is merged without this will it just gracefully not receive pan/zoom events?

**Comment by moffatman** [comment]
@justinmc Framework PR needs to be merged first as merging only engine changes will break scrolling. If only the framework PR is merged there should be no changes to existing behaviour. In fact the Web engine will never get a PR, as it's not possible to get trackpad gestures to work there. So there is no rush or synchronization needed in merging the engine PRs after the framework lands. This specific one I have been slow on amending since I can't build for android on my ARM mac.

**Comment by justinmc** [comment]
Sounds good, thanks for the clarification.

**Comment by moffatman** [comment]
@blasten 
I want to try making the following change to robolectric to fix the failure, can't seem to get the tests to use a locally-built version though.
```
diff --git a/shadows/framework/src/main/java/org/robolectric/shadows/ShadowMotionEvent.java b/shadows/framework/src/main/java/org/robolectric/shadows/ShadowMotionEvent.java
index 1ee385479..682938376 100644
--- a/shadows/framework/src/main/java/org/robolectric/shadows/ShadowMotionEvent.java
+++ b/shadows/framework/src/main/java/org/robolectric/shadows/ShadowMotionEvent.java
@@ -52,7 +52,7 @@ import org.robolectric.util.ReflectionHelpers;
  */
 @SuppressWarnings({"UnusedDeclaration"})
 @Implements(MotionEvent.class)
-public class ShadowMotionEvent {
+public class ShadowMotionEvent extends ShadowInputEvent {
 
   private static NativeObjRegistry<NativeInput.MotionEvent> nativeMotionEventRegistry =
       new NativeObjRegistry<>(NativeInput.MotionEvent.class);
```

**Comment by blasten** [comment]
@moffatman that's weird. I think the workaround requires to supply a shadow implementation that extends [InputEvent](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/view/InputEvent.java;drc=7bf53f9cd0a49bc13e7b2477f7ac6397a9296fa4;l=28). 

For more, see http://robolectric.org/extending/

**Comment by moffatman** [comment]
@blasten Making a new shadow implementation will be a lot of code since we cannot extend the existing one as we wish to change its parent class. I updated the test to just use Mockito mock instead of shadowed object, and it's running fine now.

**Comment by chinmaygarde** [comment]
Ping @blasten

File: shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java
Comment (dkwingsmt) [inline shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java:254]: Extract this recurring condition as a boolean.

Comment (dkwingsmt) [reply shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java:254]: Or do you think it makes sense to extract the trackpad related data to another function? Because it seems that there are  contains exclusive logic on both side, such as adding to and removing from `ongoingPans`, and many `if` clauses.

If we define a "data packet" class (as commented by a TODO on L207 of this file), then we don't have to worry about assignment order or packet structure. We can assign the common fields first, then split exclusive parts into two functions.

**Review by dkwingsmt — CHANGES_REQUESTED** [review]

File: shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java
Comment (moffatman) [reply shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java]: Just extracted the condition for now. It would be nice to use a data packet if it could be used in the test class as well. I don't know that much about how the java code interoperates with the rest of the engine, but I assume we can't change the signature on FlutterRenderer::dispatchPointerDataPacket ?

Comment (dkwingsmt) [reply shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java]: The "data packet" class is actually just a builder. Check out https://github.com/flutter/engine/blob/main/lib/ui/window/pointer_data_packet.h .

**Comment by chinmaygarde** [comment]
@dkwingsmt I think this is good for another review. Thanks.

File: shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java
Comment (dkwingsmt) [inline shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java:234]: Add comment:

Some implementations translates trackpad scrolling into a mouse down-move-up event sequence with buttons: 0, such as ARC on a Chromebook. See https://github.com/flutter/engine/pull/11420, a legacy implementation that uses the same condition but converts differently.

**Review by dkwingsmt — APPROVED** [review]
LGTM except for a comment

**Comment by dkwingsmt** [comment]
@blasten Would you like to make a second review?

File: shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java
Comment (dkwingsmt) [inline shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java:238]:
```suggestion
        // sequence with buttons: 0, such as ARC on a Chromebook. See https://github.com/flutter/engine/pull/11420, a legacy
```

File: shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java
Comment (dkwingsmt) [inline shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java:241]:
```suggestion
    if (pointerKind == PointerDeviceKind.MOUSE) {
      buttons = event.getButtonState() & 0x1F;
    } else if (pointerKind == PointerDeviceKind.TOUCH) {
      buttons = 0;
      if (event.getButtonState() == 0
          && event.getSource() == InputDevice.SOURCE_MOUSE
          && pointerChange == PointerChange.DOWN) {
        // Chromebook with ARC translates trackpad scrolling into a touch down-move-up event
        // sequence, which can be identified by a distinctive combination of input device "mouse",
        // tool type "finger", and buttons 0.
        ongoingPans.put(event.getPointerId(pointerIndex), viewToScreenCoords);
      }
```

Comment (dkwingsmt) [reply shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java:241]: Can you also add a test case for this special case?

Comment (moffatman) [reply shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java:241]: Any concern that we have no idea which proportion of devices this will work on? I guess it's based on undocumented behaviour anyway, and Chrome OS is always up to date.

Comment (dkwingsmt) [reply shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java:241]: Also maybe it's better to use `switch` instead.

Comment (dkwingsmt) [reply shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java:241]: Never mind, I'll handle this change in another PR. The current code passes all tests anyway.

**Comment by dkwingsmt** [comment]
@blasten Would you like to make a second review of this PR?

File: shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java
Comment (blasten) [inline shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java:237]: can `_POINTER_BUTTON_PRIMARY` be removed?  Do we know why this was introduced in #11420?

Comment (blasten) [reply shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java:237]: @dkwingsmt

File: shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java
Comment (blasten) [inline shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java:72]: can this be package only? also, it's missing the annotation `@VisibleForTesting`.

File: shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java
Comment (dkwingsmt) [reply shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java:237]: Yes, we should remove this. It has no more uses.

**Comment by blasten** [comment]
This change LGTM once the two minor comments are resolved

**Comment by zanderso** [comment]
@moffatman From PR review triage: it looks like the next step for this PR is to respond to the comments from @blasten

**Comment by chinmaygarde** [comment]
@blasten is OOO for the rest of this week.

**Review by blasten — APPROVED** [review]
LGTM

**Pull request merged**
